File: sdk/lib/rust/eruption-rs/src/connection.rs
/*  SPDX-License-Identifier: LGPL-3.0-or-later  */

/*
    This file is part of the Eruption SDK.

    The Eruption SDK is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    The Eruption SDK is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with the Eruption SDK.  If not, see <http://www.gnu.org/licenses/>.

    Copyright (c) 2019-2023, The Eruption Development Team
*/

use crate::canvas::Canvas;
use crate::hardware::HotplugInfo;
use crate::transport::{LocalTransport, ServerStatus, Transport};
use crate::Result;
use std::sync::Arc;
use tracing_mutex::stdsync::RwLock;

#[derive(Clone)]
pub struct Connection {
    con: Arc<RwLock<Box<dyn Transport + Sync + Send>>>,
}

impl Connection {
    pub fn new(connection_type: ConnectionType) -> Result<Self> {
        Ok(Self {
            con: Arc::new(RwLock::new(make_transport(&connection_type)?)),
        })
    }

    pub fn connect(&self) -> Result<()> {
        self.con.write().unwrap().connect()
    }

    pub fn disconnect(&self) -> Result<()> {
        self.con.write().unwrap().disconnect()
    }

    pub fn submit_canvas(&self, canvas: &Canvas) -> Result<()> {
        self.con.read().unwrap().submit_canvas(canvas)
    }

    pub fn get_canvas(&self) -> Result<Canvas> {
        self.con.read().unwrap().get_canvas()
    }

    pub fn get_server_status(&self) -> Result<ServerStatus> {
        self.con.read().unwrap().get_server_status()
    }

    pub fn notify_device_hotplug(&self, hotplug_info: &HotplugInfo) -> Result<()> {
        self.con.read().unwrap().notify_device_hotplug(hotplug_info)
    }

    pub fn notify_resume_from_suspend(&self) -> Result<()> {
        self.con.read().unwrap().notify_resume_from_suspend()
    }
}

impl Drop for Connection {
    fn drop(&mut self) {
        let _ = self.disconnect();
    }
}

/// The type of the connection
#[derive(Debug, Clone)]
pub enum ConnectionType {
    /// Unknown connection type
    Unknown,

    /// Local transport
    Local,

    /// Type REMOTE is currently not implemented
    Remote,
}

fn make_transport(_connection_type: &ConnectionType) -> Result<Box<impl Transport>> {
    Ok(Box::new(LocalTransport::new()?))
}

File: sdk/lib/python/eruption/connection.py
#    This file is part of Eruption.
#
#    Eruption is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    Eruption is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with Eruption.  If not, see <http://www.gnu.org/licenses/>.
#
#    Copyright (c) 2019-2022, The Eruption Development Team


from eruption.transport.local import LocalTransport


class Connection:
    """Connection to a running instance of the Eruption daemon"""

    # Connection types
    UNKNOWN: int = 0  # unknown connection
    LOCAL: int = 1  # local transport
    REMOTE: int = 2  # type REMOTE is currently not implemented

    def __init__(self, *args, **kwargs):
        if not ("type" in kwargs and kwargs["type"] == Connection.LOCAL):
            raise InvalidParam("Invalid or unsupported connection type")
        else:
            self.connection_type = kwargs["type"]

    def connect(self):
        """Connect to a running instance of Eruption"""
        self._con = LocalTransport()
        self._con.connect()

        self.connected = True

    def disconnect(self):
        """Disconnect from Eruption"""
        if not self.is_connected():
            raise NotConnectedError("Not connected")

        self._con = None
        self.connected = False

    def is_connected(self):
        """Get connection state, returns True if we are connected to a running Eruption
           instance, otherwise returns False"""
        return self.connected

    def get_server_status(self):
        """Get the status of a running Eruption instance"""
        if not self.is_connected():
            raise NotConnectedError("Not connected")

        result = self._con.get_server_status()
        return result

    def get_active_profile(self):
        """Get the file path of the active profile"""
        if not self.is_connected():
            raise NotConnectedError("Not connected")

        result = self._con.get_active_profile()
        return result

    def switch_profile(self, profile_file):
        """Switches the active profile to one given in the file path"""
        if not self.is_connected():
            raise NotConnectedError("Not connected")

        result = self._con.switch_profile(profile_file)
        return result

    def set_parameters(self, profile_file, script_file, **parameters):
        """Set parameters for the given profile's script"""
        if not self.is_connected():
            raise NotConnectedError("Not connected")

        result = self._con.set_parameters(profile_file, script_file, **parameters)
        return result

    def submit_canvas(self, canvas, *args, **kwargs):
        """Submit the canvas to Eruption for realization"""
        if not self.is_connected():
            raise NotConnectedError("Not connected")

        result = self._con.submit_canvas(canvas)
        return result

    def notify_device_hotplug(self, hotplug_info, *args, **kwargs):
        """Notify Eruption about a device hotplug event"""
        if not self.is_connected():
            raise NotConnectedError("Not connected")

        result = self._con.notify_device_hotplug(hotplug_info)
        return result


class InvalidParam(Exception):
    pass


class NotConnectedError(Exception):
    pass


class ConnectionFailed(Exception):
    pass
